### PR TITLE
[B-803] Structure work items

### PR DESCRIPTION
**Pull request by AnthonyBuisset**
closes B-803

### CODE REVIEW

File: migrations/2023-03-21-141824_create-work-items-table/up.sql [hunk 1]
```diff
@@ -0,0 +1,28 @@
...
+INSERT INTO work_items
+SELECT
+	payment_id,
+	work_item [1] AS repo_owner,
+	work_item [2] AS repo_name,
+	work_item [4]::bigint AS issue_number
+FROM (
+	SELECT
+		payment_id,
+		regexp_match (work_item::text,
```
Comment (ofux) [inline migrations/2023-03-21-141824_create-work-items-table/up.sql:18]: I love postgres

File: backend/event-listeners/src/domain/projections/work_item.rs [hunk 2]
```diff
@@ -11,3 +13,24 @@ pub struct WorkItem {
...
 }
+
+#[derive(Error, Debug)]
+pub enum Error {
+	#[error("Invalid work item url")]
+	InvalidUrl,
+}
+
+impl WorkItem {
+	pub fn from_url(payment_id: PaymentId, url: Url) -> Result<Self, Error> {
```
Comment (ofux) [inline backend/event-listeners/src/domain/projections/work_item.rs:24]: I thin a few unit tests would be cool for this kind of method

File: backend/event-listeners/src/domain/projectors/budget.rs [hunk 3]
```diff
@@ -73,6 +82,22 @@ impl EventListener for Projector {
 							*requested_at,
 							None,
 						))?;
+
+						serde_json::from_value::<Reason>(reason.clone())
+							.map_err(|e| SubscriberCallbackError::Fatal(anyhow!(e)))?
```
Comment (ofux) [inline backend/event-listeners/src/domain/projectors/budget.rs:87]: `SubscriberCallbackError::Discard`

File: backend/event-listeners/src/domain/projectors/budget.rs [hunk 4]
```diff
@@ -73,6 +82,22 @@ impl EventListener for Projector {
...
+							.work_items
+							.iter()
+							.try_for_each(|url| {
+								let url = url.parse().map_err(|e: url::ParseError| {
+									SubscriberCallbackError::Fatal(anyhow!(e))
```
Comment (ofux) [inline backend/event-listeners/src/domain/projectors/budget.rs:92]: `SubscriberCallbackError::Discard`

File: backend/event-listeners/src/domain/projectors/budget.rs [hunk 5]
```diff
@@ -73,6 +82,22 @@ impl EventListener for Projector {
...
+								})?;
+
+								self.work_item_repository
+									.upsert(&WorkItem::from_url(*payment_id, url)?)?;
```
Comment (ofux) [inline backend/event-listeners/src/domain/projectors/budget.rs:96]: Extract `&WorkItem::from_url(*payment_id, url)` and map error to `SubscriberCallbackError::Discard`. 
Error coming from the `upsert` must still be mapped to `SubscriberCallbackError::Fatal`

**Review by ofux — CHANGES_REQUESTED** [review]
LGTM except the mapping with `SubscriberCallbackError`.
BTW, some errors in the same file that are unrelated to this PR seem to be badly mapped as well.

File: backend/event-listeners/src/domain/projections/work_item.rs (hunk 2 above)
Comment (AnthonyBuisset) [reply backend/event-listeners/src/domain/projections/work_item.rs:24]: indeed

File: backend/event-listeners/src/domain/projectors/budget.rs (hunk 5 above)
Comment (AnthonyBuisset) [reply backend/event-listeners/src/domain/projectors/budget.rs:96]: done, as well as for the amount checking

**Review by ofux — APPROVED** [review]

**Pull request merged**